src/segment-loader.js

```diff
@@ -1178,6 +1221,187 @@ export default class SegmentLoader extends videojs.EventTarget {
                                      this.handleUpdateEnd_.bind(this));
   }
 
+  /**
+   * Adjust forward and backward buffers based on amount of time and bytes
+   * currently in the buffer
+   *
+   * @private
+   *
+   * @param {Boolean=} setByteLimit
+   *                   Whether we should set maxBytes or just the buffer sizes
+   */
+  adjustBuffers_(setByteLimit) {
+    if (this.mediaRequests === 1) {
+      // we failed on the first request, so don't adjust buffers
+      return;
+    }
+    this.removeNonBufferedSegmentCues();
+    let buffered = this.buffered_();
+    let forwardBuffer = this.maxGoalBufferLength_();
+    let reverseBuffer = this.backBufferLength_();
+    let configTotalBuffer = forwardBuffer + reverseBuffer;
+    let bytesInBuffer = 0;
+    let secondsInBuffer = 0;
+    let maxForwardBuffer;
+    let maxReverseBuffer;
+    let i;
+    let loggerMsg;
+
+    bytesInBuffer = this.minBufferedBytes();
+    for (i = 0; i < buffered.length; i++) {
+      secondsInBuffer += buffered.end(i) - buffered.start(i);
+    }
+    if (!this.playlist_.attributes || !this.playlist_.attributes.BANDWIDTH) {
+      // we're downloading a single rate (and don't know what that rate is)
+      // so we just use the seconds in the buffer to calculate the new buffer
+      // sizes.
+      maxForwardBuffer = Math.max((forwardBuffer / configTotalBuffer) * secondsInBuffer,
+        this.playlist_.targetDuration);
+      maxReverseBuffer = secondsInBuffer - maxForwardBuffer;
+    } else {
+      // since we potentially have more than one rate in the buffer, we use
+      // the amount of bytes in the buffer to calculate the upper limit on
+      // the equivalent time at the current rate
+      let upperTimeLimit = (bytesInBuffer * 8) / this.playlist_.attributes.BANDWIDTH;
+
+      maxForwardBuffer = (forwardBuffer / configTotalBuffer) * upperTimeLimit;
+      maxReverseBuffer = (reverseBuffer / configTotalBuffer) * upperTimeLimit;
+    }
+
+    maxForwardBuffer = Math.min(forwardBuffer, maxForwardBuffer);
+    maxReverseBuffer = Math.min(reverseBuffer, maxReverseBuffer);
+
+    loggerMsg = '\n\tbytes in buffer: ' + bytesInBuffer +
+      ', seconds in buffer: ' + secondsInBuffer +
+      '\n\tforward buffer: ' + forwardBuffer + ' => ' + maxForwardBuffer +
+      '\n\treverse buffer: ' + reverseBuffer + ' => ' + maxReverseBuffer;
+
+    if (setByteLimit) {
+      loggerMsg += '\n\tbyte limit: ' + this.maxBytes_ + ' => ' + bytesInBuffer;
+      this.maxBytes_ = Math.min(this.maxBytes_, bytesInBuffer);
+    }
+    this.logger_('adjustBuffers_', loggerMsg);
+
+    this.goalBufferLength_(maxForwardBuffer);
```

Comment:
yeah, or just plain `forwardBufferLength`

Reply:
Do you think `forwardBufferGoal` is too ambiguous? I guess `forwardBufferGoalSeconds` would be super explicit about what it is.

Reply:
My question to answer your question would be: is that the number of seconds we currently have buffered or what we're getting close to?

Reply:
Also note to self: I should double check we can actually rename this since it's publicly facing

Reply:
It's the target for the number of seconds we want in the buffer.

If it's really causing that much confusion, I think we should just rename it and have `goalBufferLength` throw a deprecation notice and call the new name.

Reply:
🤔 I do think we want to avoid making breaking changes if we can, but a deprecation warning might be a good compromise. Definitely will talk to the rest of the team soon.